### PR TITLE
Backend refactor

### PR DESCRIPTION
**Pull request by RidhaChowdhury**
The backend has a router-service-based architecture that makes the main very minimal and allows us to expand the API by adding files.

### CODE REVIEW

File: backend/src/settings.py [hunk 1]
```diff
@@ -0,0 +1,4 @@
+origins = [
+    "http://localhost:5173",
+    "chrome-extension://joecpfmckhoobfipcmmhhkkmohhjebmm"
```
Comment (ehedges03) [inline backend/src/settings.py:3]: Does this ID change?

File: backend/src/routers/__init__.py
Comment (ehedges03) [inline backend/src/routers/__init__.py:1]: Is this file required? It does not look to be doing anything.

File: backend/src/services/langchain_service.py [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+hint_prompt_template = '''...'''
+hint_prompt = PromptTemplate.from_template(template=hint_prompt_template)
+
+def get_llm() -> OpenAI:
+    openai_api_key = os.getenv("OPENAI_API_KEY")
+    return OpenAI(openai_api_key=openai_api_key, temperature=0.5, max_tokens=1024)
+    
+def get_hint_chain(description: str, code: str) -> str:
+    llm = get_llm()
+    hint_chain = LLMChain(llm=llm, prompt=hint_prompt)
```
Comment (ehedges03) [inline backend/src/services/langchain_service.py:14]: This is meaningless at the moment, but do you have to call OpenAI() and LLMChain() for every request? I assume OpenAI forms the connection and LLMChain just sets up the ability to chain prompts together with the initial prompt being `hint_prompt`.

**Review by ehedges03 — COMMENTED** [review]
As far as logic goes it lgtm.

File: backend/main.py [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+
+app.add_middleware(
+    CORSMiddleware,
+    allow_origins=origins,
+    allow_credentials=True,
+    allow_methods=["*"],
+    allow_headers=["*"]
+)
+
+app.include_router(hint_router)
```
Comment (aahmad4) [inline backend/main.py:17]: nit: Add newline

File: backend/src/routers/__init__.py
Comment (aahmad4) [reply backend/src/routers/__init__.py:1]: Might be a placeholder for when we resume work on that hint endpoint

File: backend/src/services/langchain_service.py (hunk 2 above)
Comment (aahmad4) [reply backend/src/services/langchain_service.py:14]: I have a similar question - is there a way we can establish that connection to OpenAI when the server first starts up? Then keep re-using that connection for all hint requests

File: backend/src/settings.py [hunk 4]
```diff
@@ -0,0 +1,4 @@
...
+]
```
Comment (aahmad4) [inline backend/src/settings.py:4]: nit: Add newline

File: backend/src/settings.py [hunk 5]
```diff
@@ -0,0 +1,4 @@
+origins = [
```
Comment (aahmad4) [inline backend/src/settings.py:1]: What is this file for out of curiosity?

File: .vscode/launch.json [hunk 6]
```diff
@@ -0,0 +1,40 @@
+{
```
Comment (aahmad4) [inline .vscode/launch.json:1]: What is this file for out of curiosity? Does it add some abstractions which will make the project easier to run?

File: backend/src/services/langchain_service.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+from langchain import LLMChain, PromptTemplate
+
+hint_prompt_template = '''...'''
+hint_prompt = PromptTemplate.from_template(template=hint_prompt_template)
+
+def get_llm() -> OpenAI:
+    openai_api_key = os.getenv("OPENAI_API_KEY")
+    return OpenAI(openai_api_key=openai_api_key, temperature=0.5, max_tokens=1024)
+    
+def get_hint_chain(description: str, code: str) -> str:
```
Comment (aahmad4) [inline backend/src/services/langchain_service.py:12]: Can we just take in a `HintRequest` object here? Rather than taking in the description and code separately

File: backend/src/routers/hint_router.py [hunk 8]
```diff
@@ -0,0 +1,14 @@
...
+from ..services.langchain_service import get_hint_chain
+
+hint_router = APIRouter()
+
+@hint_router.get("/")
+async def root() -> str:
+    return "Welcome to LeetCopilot!"
+
+@hint_router.post("/hint", response_model=str)
+async def get_hint(promptData: HintRequest):
```
Comment (aahmad4) [inline backend/src/routers/hint_router.py:12]: Should we be setting the return type of this to `HintResponse`?

File: backend/src/services/langchain_service.py (hunk 2 above)
Comment (aahmad4) [inline backend/src/services/langchain_service.py:14]: Do we need to update `hint_prompt` first to actually include the code and problem description? Is this just temporary?

File: backend/src/services/langchain_service.py [hunk 9]
```diff
@@ -0,0 +1,16 @@
...
+    llm = get_llm()
+    hint_chain = LLMChain(llm=llm, prompt=hint_prompt)
+    hint = hint_chain.run(description=description, code=code).strip()
```
Comment (aahmad4) [inline backend/src/services/langchain_service.py:15]: What exactly does this part do? Does that `run` method take in any type of data, and what is it for?

File: backend/src/settings.py (hunk 1 above)
Comment (RidhaChowdhury) [reply backend/src/settings.py:3]: Shouldnt

File: backend/src/services/langchain_service.py (hunk 2 above)
Comment (RidhaChowdhury) [reply backend/src/services/langchain_service.py:14]: Yeah we can attach it to the fastAPI app instance.

File: .vscode/launch.json (hunk 6 above)
Comment (RidhaChowdhury) [reply .vscode/launch.json:1]: Yeah in vscode you can use the debugger and start the frontend, backend, or the full stack with one click

File: backend/src/services/langchain_service.py (hunk 2 above)
Comment (RidhaChowdhury) [reply backend/src/services/langchain_service.py:14]: I think we can do that when we migrate the LLM fully to llama in a separate PR

File: backend/src/settings.py (hunk 5 above)
Comment (RidhaChowdhury) [reply backend/src/settings.py:1]: This is where we store data that is relevant to the backend instance such as acceptable origins.

File: backend/src/services/langchain_service.py (hunk 7 above)
Comment (RidhaChowdhury) [reply backend/src/services/langchain_service.py:12]: Oh for sure.

File: backend/src/routers/hint_router.py (hunk 8 above)
Comment (RidhaChowdhury) [reply backend/src/routers/hint_router.py:12]: Yes

File: backend/src/services/langchain_service.py (hunk 2 above)
Comment (RidhaChowdhury) [reply backend/src/services/langchain_service.py:14]: The problem description and code are added when you run the chain.

File: backend/src/services/langchain_service.py (hunk 9 above)
Comment (RidhaChowdhury) [reply backend/src/services/langchain_service.py:15]: It takes the llm, the chain, and the inputs and runs the llm query

**Comment by RidhaChowdhury** [comment]
closes #27

**Comment by RidhaChowdhury** [comment]
closes #26
